[SPARK-2774] Set preferred locations for reduce tasks

shivaram opened the pull request:
Set preferred locations for reduce tasks. 
The basic design is that we maintain a map from reducerId to a list of (sizes, locations) for each shuffle. We then pick the machines with the top 5 sizes and set that as preferred locations. 

Picking the top5 needs sorting (well its a little cheaper than sorting, but still), so we restrict this feature to cases where we have fewer than 1000 map tasks and 1000 reduce tasks.

Resubmission of #4576 and  #1697. Hopefully third time is a charm ? cc @kayousterhout for review

shivaram committed:
Set preferred locations for reduce tasks
This is another attempt at #1697 addressing some of the earlier concerns.
This adds a couple of thresholds based on number map and reduce tasks
beyond which we don't use preferred locations for reduce tasks.

This patch also fixes some bugs in DAGSchedulerSuite where the MapStatus
objects created didn't have the right number of reducers set.

shivaram committed:
Fix style issues

shivaram committed:
Fix bug introduced by line length adjustment

shivaram committed:
Merge branch 'master' of https://github.com/apache/spark into reduce-locations

shivaram committed:
Merge branch 'master' of https://github.com/apache/spark into reduce-locations

shivaram committed:
Merge branch 'master' of https://github.com/apache/spark into reduce-locations

shivaram committed:
Merge branch 'master' of https://github.com/apache/spark into reduce-locations

shivaram committed:
Merge branch 'master' of https://github.com/apache/spark into reduce-locations

Conflicts:
	core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala
	core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

shivaram committed:
Fix merge issues

JoshRosen commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,32 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+  // The map is keyed by reducerId and for each reducer the value contains the array
+  // of (location, size) of map outputs.
+  //
+  // This method is not thread-safe
```
Why not use the Scaladoc syntax for this?

shivaram replied:
Now using scaladoc syntax

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34210 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34210/consoleFull) for   PR 6652 at commit [`e7d5449`](https://github.com/apache/spark/commit/e7d5449238a6f674ab4500c01ac0be0bffd53832).

JoshRosen commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,32 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+    : Option[Map[Int, Array[(BlockManagerId, Long)]]] = {
+    if (!statusByReducer.contains(shuffleId) && mapStatuses.contains(shuffleId)) {
+      val statuses = mapStatuses(shuffleId)
+      if (statuses.length > 0) {
```
`statuses.nonEmpty` ?

shivaram replied:
Done

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1395,6 +1404,23 @@ class DAGScheduler(
+              // Select first few locations as preferred locations for the reducer
+              val topLocs = CollectionUtils.takeOrdered(
+                status(partition).iterator, NUM_REDUCER_PREF_LOCS)(
+                  Ordering.by[(BlockManagerId, Long), Long](_._2).reverse).toSeq
```
It looks like this `Ordering.by` is effectively a constant?  Maybe we can pull it out of the `map` here and reuse it for all loop iterations.

shivaram replied:
I've moved this into MapOutputTracker now and I've pulled it out of the loop there

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1395,6 +1404,23 @@ class DAGScheduler(
+              val topLocs = CollectionUtils.takeOrdered(
+                status(partition).iterator, NUM_REDUCER_PREF_LOCS)(
+                  Ordering.by[(BlockManagerId, Long), Long](_._2).reverse).toSeq
+              return topLocs.map(_._1).map(loc => TaskLocation(loc.host, loc.executorId))
```
It looks like you're `return`ing from a `map` here?  Maybe we can replace this map with a `while` loop instead.

shivaram replied:
This is now avoided as we return just one reducer's information from the MapOutputTracker. So we have an `if(option.nonEmpty) return` here now instead of `map`

JoshRosen commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -137,6 +138,14 @@ class DAGScheduler(
 
+  // Number of map, reduce tasks above which we do not assign preferred locations
+  // based on map output sizes.
+  private val SHUFFLE_PREF_MAP_THRESHOLD = 1000
```
This could be `private[this]` instead, but not a big deal.

kayousterhout replied:
What's the difference? (Have always thought they were the same!)

shivaram replied:
http://mail-archives.apache.org/mod_mbox/incubator-spark-dev/201402.mbox/%3CCAAsvFPk5rZfJv=Qr+Dufs=iKVvyrQmgPJY+pS_yTciuPeJvJLA@mail.gmail.com%3E explains it better than I can

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -21,7 +21,7 @@ import java.io._
 import java.util.zip.{GZIPInputStream, GZIPOutputStream}
 
-import scala.collection.mutable.{HashSet, Map}
+import scala.collection.mutable.{HashSet, Map, HashMap}
```
import ordering

shivaram replied:
Done

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -232,6 +232,12 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+  private val statusByReducer =
+    new TimeStampedHashMap[Int, HashMap[Int, Array[(BlockManagerId, Long)]]]()
+
+
```
nit: extra newline

shivaram replied:
Fixed

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1395,6 +1404,23 @@ class DAGScheduler(
             return locs
           }
         }
+      case s: ShuffleDependency[_, _, _] =>
```
Can you add a high-level comment here explaining what this is doing? Something like "For reduce tasks, return the 5 locations with the largest map outputs as preferred locations"

shivaram replied:
Added -- Also I moved the comment for narrow dependencies close to that.

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -137,6 +138,14 @@ class DAGScheduler(
   private[scheduler] val eventProcessLoop = new DAGSchedulerEventProcessLoop(this)
   taskScheduler.setDAGScheduler(this)
 
+  // Number of map, reduce tasks above which we do not assign preferred locations
```
Could you add a comment here saying that we limit the size because of scalability issues with sorting the best locations?

shivaram replied:
Done

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -232,6 +232,12 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
 
+  // For each shuffleId we also maintain a Map from reducerId -> (location, size)
+  // Lazily populated whenever the statuses are requested from DAGScheduler
+  private val statusByReducer =
```
nit: could you rename this to shuffleIdToReduceLocations? With the current name, I expected the main map to be keyed on reduce id.

shivaram replied:
Done

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,32 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+  // This method is not thread-safe
+  def getStatusByReducer(
+      shuffleId: Int,
+      numReducers: Int)
```
As discussed offline, take in a parameter here for maxNumberLocations, and then do the filtering for top 5 on line 309-ish when you create the status method (this avoids some of the scalability issues of creating a gigantic map)

kayousterhout replied:
Also, what about changing this to something like "getMapLocationsWithBiggestOutputs", and having it also accept the reduceId (and then return Option[Array[(BlockManagerId, Long)]], since the caller always just takes one element out of the map? And then you can just add a comment internally saying that you populate the whole map for use later.

shivaram replied:
I've moved the `takeOrdered` to MapOutputTracker now. Unfortunately some of the savings in constructing the array can't be done as we need the groupBy here (for the comment below).
Also I'm hoping the `groupBy` is not expensive if we have say a 1000 elements in the array. 

@JoshRosen let me know if there is a better way to do this.

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,32 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+    if (!statusByReducer.contains(shuffleId) && mapStatuses.contains(shuffleId)) {
+      val statuses = mapStatuses(shuffleId)
+      if (statuses.length > 0) {
+        statusByReducer(shuffleId) = new HashMap[Int, Array[(BlockManagerId, Long)]]
```
Also is it possible to combine entries for the same block manager here? I think that, right now, if multiple maps are at the same block manager, they won't be aggregated when thinking about which block manager ID has the most output, so if you have

BM 1: 2 bytes
BM 1: 2 bytes
BM 2: 3 bytes,

right now BM2 will be considered the biggest.

shivaram replied:
Great point. I didn't take this into account before. Now added a test in MapOutputTrackerSuite to test this.

SparkQA commented:
[Test build #34210 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34210/consoleFull) for   PR 6652 at commit [`e7d5449`](https://github.com/apache/spark/commit/e7d5449238a6f674ab4500c01ac0be0bffd53832).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected class CodeGenContext `
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `abstract class Predicate `
  * `      class SpecificPredicate extends $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test PASSed.

shivaram committed:
Address code review comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34220 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34220/consoleFull) for   PR 6652 at commit [`0df3180`](https://github.com/apache/spark/commit/0df31800d6a633e829f0ceb8cf2229e9e9fd940b).

SparkQA commented:
[Test build #34220 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34220/consoleFull) for   PR 6652 at commit [`0df3180`](https://github.com/apache/spark/commit/0df31800d6a633e829f0ceb8cf2229e9e9fd940b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram committed:
Fix style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34222 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34222/consoleFull) for   PR 6652 at commit [`8e31266`](https://github.com/apache/spark/commit/8e312661361738be13f71134ea69555db2ae1b60).

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,41 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+          val locs = statuses.map { s =>
+            (s.location, s.getSizeForBlock(r))
+          }.groupBy(_._1).mapValues { sizes =>
+            sizes.map(_._2).reduceLeft(_ + _)
```
is there a reason to not use "sizes.map(_._2).sum" instead here? (just slightly easier to read)

kayousterhout replied:
Actually what about:

val locs = statuses.groupBy(s.location).mapValues { statusesAtLocation =>
  statusesAtLocation.map(_.getSizeForBlock(r)).sum }

shivaram replied:
I made this imperative now per Josh's comment below. I think the new version is pretty readable, but let me now if you think its good

JoshRosen commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,41 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+        var r = 0
+        while (r < numReducers) {
+          // Add up sizes of all blocks at the same location
+          val locs = statuses.map { s =>
```
If perf. is important here, why not write some java-style code that does an aggregate-by-key using a mutable hashmap?

JoshRosen replied:
Actually, I think we should just go with a straightforward imperative code here.  I don't think it will be significantly longer and can't be slow by accident (it'll also create less garbage).

shivaram replied:
You are right. Its the same 3 lines of code imperatively as well. And create a bit less garbage I reuse the hashmap across reducers now.  Let me know if this looks okay

kayousterhout commented on core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala:
```diff
@@ -205,4 +205,29 @@ class MapOutputTrackerSuite extends SparkFunSuite {
+    // Setup 3 map tasks
+    // on hostA with output size 1
+    // on hostA with output size 1
+    // on hostB with output size 2
```
Can you do 2, 2, 3 as the sizes here? Otherwise this seems non-deterministic, since the two machines have equal sizes of data?

shivaram replied:
Oh yes - sorry this should have been 2,2,3. Made it now

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -137,6 +137,15 @@ class DAGScheduler(
+  // NOTE: This should be less than 2000 as we use HighlyCompressedMapStatus beyond that
+  private[this] val SHUFFLE_PREF_REDUCE_THRESHOLD = 1000
+  // Number of preferred locations to use for reducer tasks
+  private[scheduler] val NUM_REDUCER_PREF_LOCS = 5
```
one more comment request here: can you add one sentence explaining this? something like "making this smaller will focus on the locations where the most data can be read locally, but may lead to lower scheduling efficiency or the delay scheduling timers expiring if all of those locations are busy"

shivaram replied:
Added now

kayousterhout commented:
Ok this looks mostly good, subject to @JoshRosen's comments about scalability -- just a few last comments to improve readability.

SparkQA commented:
[Test build #34222 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34222/consoleFull) for   PR 6652 at commit [`8e31266`](https://github.com/apache/spark/commit/8e312661361738be13f71134ea69555db2ae1b60).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

shivaram committed:
Address some more comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34239 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34239/consoleFull) for   PR 6652 at commit [`9d5831a`](https://github.com/apache/spark/commit/9d5831a9ff918f1e7886aae0a9f0dd4d732303c0).

SparkQA commented:
[Test build #34239 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34239/consoleFull) for   PR 6652 at commit [`9d5831a`](https://github.com/apache/spark/commit/9d5831a9ff918f1e7886aae0a9f0dd4d732303c0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public final class UnsafeRow extends BaseMutableRow `
  * `public abstract class BaseMutableRow extends BaseRow implements MutableRow `
  * `public abstract class BaseRow implements Row `
  * `  protected class CodeGenContext `
  * `abstract class BaseMutableProjection extends MutableProjection `
  * `      class SpecificProjection extends $`
  * `class BaseOrdering extends Ordering[Row] `
  * `      class SpecificOrdering extends $`
  * `abstract class Predicate `
  * `      class SpecificPredicate extends $`
  * `abstract class BaseProject extends Projection `
  * `    class SpecificProjection extends $`
  * `    final class SpecificRow extends $`

AmplabJenkins commented:
Merged build finished. Test PASSed.

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,44 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+        while (r < numReducers) {
+          var i = 0
+          locs.clear()
+          while (i < statuses.length) {
```
just a tiny style nit here: is there a reason not to write this loop as "statuses.foreach { }"? It's kind of hard to read as-is.

JoshRosen replied:
I don't think that switching to a `foreach` is necessarily the biggest readability improvement here.  Instead, I think we could rename `r` to `reduceId` and `i` to `mapId`.

shivaram replied:
Did the rename now.

kayousterhout commented:
Ok LGTM other than the small readability suggestion

ash211 commented:
@shivaram presumeably you're working on locality for reducers because it improves overall job time by decreasing the amount of shuffle bytes sent over the network.  Do you have tests showing that to be the case?

sryza commented:
When there is no skew, are there situations where this would lead to worse performance?  E.g. will it make tasks bunch up on nodes more than before and / or result in scheduling delays?

sryza commented:
Also, is there an intuitive justification for why 5 is a good number?  It seems a little weird to me that it's independent of the number of tasks, the cluster size, and the amount of skew.  Have you considered a rule like "if X% of the data falls on Y% of the nodes, mark those as preferred", where X and Y could be 80/20 or 70/30?

JoshRosen commented:
> When there is no skew, are there situations where this would lead to worse performance? E.g. will it make tasks bunch up on nodes more than before and / or result in scheduling delays?

I forget how this corner of the scheduler works, but in this case I think that we're just introducing preferences where there weren't any before, so as long as our locality delay isn't too long then I don't know that this will lead to a big queue for a particular node.

@shivaram, is there an easy way to feature-flag this?  Might be nice for being able to run benchmarks or to turn this off if it's bad for certain workloads.

shivaram commented:
@ash211 Yeah, so there are two sets of experiments that I've done which show benefits with this. 

1. The first is in the [OSDI 2014 paper](https://www.usenix.org/system/files/conference/osdi14/osdi14-paper-venkataraman.pdf). There are a bunch of things in that paper, but Figure 19 which compares Baseline to KMN-M/K=1.0 is very similar to the current patch.

2. I've run a bunch of machine learning / linear algebra workloads which use aggregation trees implemented as a series of shuffles. In these scenarios you get ~50% reduction in data sent over the network for each stage of the reduction. Here is an example of a time-breakdown diagram [before](https://www.dropbox.com/s/vjkcu3bc9cscelr/120_waterfall.pdf?dl=0) and [after](https://www.dropbox.com/s/820vlt4xuefvr7f/1016_waterfall.pdf?dl=0). In the second case here we are able to completely avoid network transfer in the second stage by aggregating across multiple partitions in the same machine.

3. @JoshRosen also included a couple of other papers which have done this in [the JIRA](https://issues.apache.org/jira/browse/SPARK-2774?focusedCommentId=14081300&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14081300)

sryza commented:
My worry is the following situation:
* Map outputs from stage 1 exhibit little skew and are distributed evenly across nodes.
* 5 nodes are chosen as preferred for stage 2.  The choice is basically random because of the lack of skew.
* Stage 2 has fewer tasks than the number of slots.
* Instead of stage 2's tasks evenly distributed across the cluster, all of them end up on the 5 nodes, which are more likely to become bottlenecked on incoming network traffic.

This isn't a disaster situation by any means.  But it also seems like it could be easily avoided with a policy that only prefers locations when it would make a difference.

sryza commented:
On the other side, I can also envision situations where it would be really helpful to request a higher number of preferred locations.  Consider a large YARN cluster where an app gets executors on a small subset of the nodes.   The app is idle for a while after running an initial job, so the initial set of executors go away.  The app wakes up later and wants to use the output of its first stage for another job.  If the scheduler could request tasks on the nodes housing the outputs of the first stage, with SPARK-4352, the ExecutorAllocationManager would be able to request executors on these nodes.

This probably makes sense as later work, but still thought it could be worth bringing up as part of the discussion.

shivaram commented:
@sryza I'm not sure why the same 5 nodes should get chosen for all the reduce tasks ? Are you concerned this will happen because we use stable sort ? We could randomize among nodes with equal size inputs I guess. Also the default locality wait is 3s, so for more non-trivial shuffles, it should fall back to the default case. 

FWIW I don't mind putting this behind a flag, especially if there are other workloads which people would want to test this with. I just want to avoid having too many flags, thresholds that make it complex to reason about what the code is doing etc.

sryza commented:
Ahhh, I misunderstood and thought that all the reduce tasks for a stage got the same locality preferences. I withdraw my concern.

shivaram committed:
Filter out zero blocks, rename variables

shivaram committed:
Add flag to turn off locality for shuffle deps

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

shivaram commented:
I briefly chatted with @kayousterhout offline as well and I added a feature flag for emergencies but kept it on by default.

@JoshRosen @sryza Will be great if you could test any workloads you have with this change and let me know if you run into problems (or if you see wins !). We are relatively early in the 1.5 cycle, so will have time to change things if we see problems.

SparkQA commented:
[Test build #34329 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34329/consoleFull) for   PR 6652 at commit [`e5d56bd`](https://github.com/apache/spark/commit/e5d56bd5fdafddd2bf0da03c7de4ff38c80e985c).

SparkQA commented:
[Test build #34329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34329/console) for   PR 6652 at commit [`e5d56bd`](https://github.com/apache/spark/commit/e5d56bd5fdafddd2bf0da03c7de4ff38c80e985c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

shivaram committed:
Merge branch 'master' of https://github.com/apache/spark into reduce-locations

shivaram committed:
Change flag name

shivaram commented:
@kayousterhout I updated the flag name to be a bit more specific. Let me know if you have any further comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34374 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34374/consoleFull) for   PR 6652 at commit [`1090b58`](https://github.com/apache/spark/commit/1090b58d2dcb286a75a3184bbf7a0232166a2941).

SparkQA commented:
[Test build #34374 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34374/console) for   PR 6652 at commit [`1090b58`](https://github.com/apache/spark/commit/1090b58d2dcb286a75a3184bbf7a0232166a2941).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram committed:
Fix line length

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34375 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34375/consoleFull) for   PR 6652 at commit [`68bc29e`](https://github.com/apache/spark/commit/68bc29e0ed9747d207e2b07ad025544d0a499d0f).

SparkQA commented:
[Test build #34375 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34375/console) for   PR 6652 at commit [`68bc29e`](https://github.com/apache/spark/commit/68bc29e0ed9747d207e2b07ad025544d0a499d0f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

kayousterhout commented:
@shivaram and I discussed this offline and whether it was possible to incorporate @sryza's point about the fact that it only makes sense to do this when there's a lot of skew in the placement of the data read by one reduce task.  We came up with a simple approach that seems strictly better than the current one (it's simpler to implement and also avoids setting preferred locations when there isn't much skew in the output). The new idea is to set the preferred locations for a reducer to be any locations that hold >20% of the data to be read by that reducer.  This will result in at most 5 locations being set (the current approach always chooses 5 locations), but more typically fewer.  20% is totally a magic number here...but intuitively it seems unnecessary to set a preferred location and add that complexity when there will be less than a 20% perf. improvement from doing so.

Unless anyone objects, @shivaram is going to push a new version that uses that approach.

shivaram committed:
Use fraction of map outputs to determine locations
Also removes caching of preferred locations to make the API cleaner

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

shivaram commented:
Thanks @kayousterhout for taking a look and proposing this new design. I agree that its more cleaner than the top-5 locations and we only get up to 5 locations at the maximum with fraction as 0.2. I've updated the code to reflect this now.

BTW I also removed caching the preferred locations in the map output tracker. This made it messy to implement an API which accepted `fractionThreshold` as an argument and I think the caching only helps in case of speculative tasks / failures (so it shouldn't hurt us much).

SparkQA commented:
[Test build #34474 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34474/consoleFull) for   PR 6652 at commit [`f5be578`](https://github.com/apache/spark/commit/f5be5784235813c0e28b13f234f89de04bc4849d).

shivaram committed:
Remove unused hash map

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,54 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+        val topLocs = locs.filter { case (loc, size) =>
+          size.toDouble / totalOutputSize >= fractionThreshold
+        }
+        // Only add this reducer to our map if we have any locations which satisfy
```
OOD comment

shivaram replied:
Fixed now

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -137,6 +137,23 @@ class DAGScheduler(
+    sc.getConf.getBoolean("spark.shuffle.reduceLocality.enabled", true)
+  // Number of map, reduce tasks above which we do not assign preferred locations
+  // based on map output sizes. We limit the size of jobs for which assign preferred locations
+  // as sorting the locations by size becomes expensive.
```
as sorting ... is OOD

shivaram replied:
Fixed now

kayousterhout commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -137,6 +137,23 @@ class DAGScheduler(
+  // location for a reduce task.
+  // Making this larger will focus on fewer locations where most data can be read locally, but
+  // may lead to more delay in scheduling if those locations are busy.
+  //
```
extra newline in the comment

shivaram replied:
Removed

SparkQA commented:
[Test build #34475 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34475/consoleFull) for   PR 6652 at commit [`897a914`](https://github.com/apache/spark/commit/897a914d56cfeb6b86c0ba75d046910455cbed7d).

kayousterhout commented on core/src/test/scala/org/apache/spark/MapOutputTrackerSuite.scala:
```diff
@@ -205,4 +205,36 @@ class MapOutputTrackerSuite extends SparkFunSuite {
+    tracker.registerMapOutput(10, 2, MapStatus(BlockManagerId("b", "hostB", 1000),
+        Array(3L)))
+
+    val topLocs50 = tracker.getLocationsWithLargestOutputs(10, 0, 1, 0.5)
```
Can you add comments to this test -- just one here saying "When the threshold is 50%, only host A should be returned a preferred location" and then below, "When the threshold is only 20%, both hosts should be returned"

shivaram replied:
Done

kayousterhout commented on core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:
```diff
@@ -800,6 +800,50 @@ class DAGSchedulerSuite
     assertDataStructuresEmpty()
   }
 
+  test("shuffle with reducer locality") {
```
ok last supernit: can you make this name a description of what should happen? "Reduce tasks should be placed locally with map output" and then for the below test "Reduce task locality preferences should only include machine with largest map output" or something?

shivaram replied:
Done

kayousterhout commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -284,6 +291,54 @@ private[spark] class MapOutputTrackerMaster(conf: SparkConf)
+        val locs = new HashMap[BlockManagerId, Long]
+        var totalOutputSize = 0L
+        var mapIdx = 0
+        while (mapIdx < statuses.length) {
```
Can you save statuses(mapIdx) here and then re-use it in the 3 places below? I just find it a little harder to read with the continued reference into the array

shivaram replied:
Good idea. Done

kayousterhout commented:
Mostly LGTM -- just a few readability comments to help whatever future poor soul wants to update the DAGScheduler code.

@sryza does this new approach adequately address your concerns?

SparkQA commented:
[Test build #34475 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34475/console) for   PR 6652 at commit [`897a914`](https://github.com/apache/spark/commit/897a914d56cfeb6b86c0ba75d046910455cbed7d).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

shivaram committed:
Address code review comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34480/consoleFull) for   PR 6652 at commit [`2ef2d39`](https://github.com/apache/spark/commit/2ef2d39344000bb1d08f37e3d889f3b8975c33c4).

SparkQA commented:
[Test build #34474 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34474/console) for   PR 6652 at commit [`f5be578`](https://github.com/apache/spark/commit/f5be5784235813c0e28b13f234f89de04bc4849d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #34480 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34480/console) for   PR 6652 at commit [`2ef2d39`](https://github.com/apache/spark/commit/2ef2d39344000bb1d08f37e3d889f3b8975c33c4).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

shivaram committed:
Remove unused import

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34519 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34519/consoleFull) for   PR 6652 at commit [`492e25e`](https://github.com/apache/spark/commit/492e25eec1722fc44d33166d7aa4fa55b7390e3d).

SparkQA commented:
[Test build #34519 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34519/console) for   PR 6652 at commit [`492e25e`](https://github.com/apache/spark/commit/492e25eec1722fc44d33166d7aa4fa55b7390e3d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1384,17 +1400,32 @@ class DAGScheduler(
+          }
+        }
+
       case _ =>
```
I think you can delete this pattern now, it shouldn't ever occur.

incidentally, this also suggests that `Depedendency` should be `sealed`.  I just tried making that change, it turned up these other warnings, which actually seem legit:

```
warn] /Users/irashid/github/spark/core/src/main/scala/org/apache/spark/rdd/CoGroupedRDD.scala:136: non-variable type argument Product2[K,Any] in type pattern org.apache.spark.OneToOneDependency[Product2[K,Any]] is unchecked since it is eliminated by erasure
[warn]       case oneToOneDependency: OneToOneDependency[Product2[K, Any]] =>
[warn]                                ^
[warn] /Users/irashid/github/spark/core/src/main/scala/org/apache/spark/rdd/CoGroupedRDD.scala:135: match may not be exhaustive.
[warn] It would fail on the following input: NarrowDependency()
[warn]     for ((dep, depNum) <- dependencies.zipWithIndex) dep match {
[warn]                                                      ^
[warn] /Users/irashid/github/spark/core/src/main/scala/org/apache/spark/rdd/SubtractedRDD.scala:109: match may not be exhaustive.
[warn] It would fail on the following input: NarrowDependency()
[warn]       dependencies(depNum) match {
[warn]                   ^
[warn] three warnings found
```

It seems that code will break if they are ever given a `RangeDependency`

squito commented:
lgtm too.  I think you could go ahead and delete the extra `case _`, though maybe it makes sense to defer changing `Dependency` to `sealed` to a separate pr.

shivaram commented:
Thanks @squito for taking a look. Yeah I think it makes to change Dependency to a sealed class but we can do it in a follow up PR. BTW in the specific case of the scheduler it might be interesting to set preferred locations for `RangeDependency` based on the size of the parent splits. I haven't run into many applications which use `RangeDependency` though.

sryza commented:
@shivaram @kayousterhout this approach addresses my concerns.  Thanks for updating!
